csup/dict.go

```diff
@@ -2,113 +2,75 @@ package csup
 
 import (
 	"io"
-	"math"
 
 	"github.com/brimdata/super"
-	"github.com/brimdata/super/zcode"
 	"golang.org/x/sync/errgroup"
 )
 
 type DictEncoder struct {
-	typ    super.Type
-	tags   map[string]byte
-	index  []byte
+	typ super.Type
+	Encoder
+
+	// These fields are derived after Encode is called.
 	counts *Uint32Encoder
-	values Encoder
+	index  []byte
+	const_ *Const
 }
 
-func NewDictEncoder(typ super.Type, values Encoder) *DictEncoder {
-	var tags map[string]byte
-	if id := typ.ID(); id != super.IDUint8 && id != super.IDInt8 && id != super.IDBool {
-		// Don't bother using a dictionary (which takes 8-bit tags) to encode
-		// other 8-bit values.
-		tags = make(map[string]byte)
+func NewDictEncoder(typ super.Type, values Encoder) Encoder {
+	if _, ok := values.(primitive); !ok {
+		panic("Dict values encoder must be a primitive")
+	}
```

Comment:
Feels off. How about this?

```suggestion
func NewDictEncoder(typ super.Type, values PrimitiveEncoder) Encoder {
```

PrimitiveEncoder would look like this:

```go
type PrimitiveEncoder interface {
	Encoder
	Dict() (Encoder, []byte, []uint32)
	ConstValue() super.Value
}
```

newPrimitiveEncoder would return that. And DictEncoder would embed it instead of Encoder.

Reply:
I can dig it

Reply:
Thanks this is better